include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -210,18 +210,21 @@ class __scan_single_wg_dynamic_kernel;
 template <typename... Name>
 class __scan_copy_single_wg_kernel;
 
+template <typename _IsGPU, typename... Name>
+class __parallel_for_kernel;
+
 //------------------------------------------------------------------------
 // parallel_for - async pattern
 //------------------------------------------------------------------------
 
 // Use the trick with incomplete type and partial specialization to deduce the kernel name
 // as the parameter pack that can be empty (for unnamed kernels) or contain exactly one
 // type (for explicitly specified name by the user)
-template <typename _KernelName>
+template <typename _IsGPU, typename _KernelName>
 struct __parallel_for_submitter;
 
 template <typename... _Name>
-struct __parallel_for_submitter<__internal::__optional_kernel_name<_Name...>>
+struct __parallel_for_submitter</*_IsGPU=*/::std::false_type, __internal::__optional_kernel_name<_Name...>>
```

Comment:
For maintainability, it would be great to just have a single implementation which has different configuration settings for GPU and CPU targets. Is this possible to do with configuration settings and the same underlying `nd_range` implementation?

Reply:
I think this is a good idea for maintainability. I will look into and see if I can find a good way to unite the two implementations with a single `nd_range` kernel.

Reply:
I have looked into this a bit more and ran some benchmarks on SPR with some different nd range kernels. I was able to tune the kernel to perform slightly better than the SYCL basic kernel by changing grain sizes, but I suppose this might be specific to SPR with the input type I used (fp32).

I think it may be the simplest and most generic solution to just leave the SYCL basic kernel for CPU devices as is and have the GPU specialization to ensure memory reads / writes are coalesced.